raftexample/wal/version.go

```diff
@@ -0,0 +1,288 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wal
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+
+	"go.etcd.io/raft/v3/raftexample/version"
+
+	"go.etcd.io/raft/v3/raftexample/etcdserverpb"
+	"go.etcd.io/raft/v3/raftexample/pbutil"
+	"go.etcd.io/raft/v3/raftpb"
+)
+
+// ReadWALVersion reads remaining entries from opened WAL and returns struct
+// that implements schema.WAL interface.
+//
+//nolint:revive
+func ReadWALVersion(w *WAL) (*walVersion, error) {
+	_, _, ents, err := w.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	return &walVersion{entries: ents}, nil
+}
+
+type walVersion struct {
+	entries []raftpb.Entry
+}
+
+// MinimalEtcdVersion returns minimal etcd able to interpret entries from  WAL log,
```

Comment:
Didn't review this in detail, but a couple of high-level comments:

- A `raft` example should probably have no knowledge about `etcd`.
- Please contain the example in a Go module, so that it's not part of the `raft` package module, and does not get imported by users.

Reply:
yes it is will WIP .. i am working on it now .. i cherrypick the commit from https://github.com/Elbehery/raft/pull/1/commits as suggested by @ahrtr  in https://github.com/etcd-io/raft/issues/2#issuecomment-1628497861

Reply:
adding the go module now

Reply:
@pav-kv added the go mod here https://github.com/etcd-io/raft/pull/153/commits/9e38675da891c8c080645bae26d92fa0e2b3f4c5